### ISSUE TITLE
Be able to get the list of keywords detected with their types

### ISSUE THREAD

**Issue by sguimont** [pegjs/pegjs#311]
I would like to be able to parse a string and gets an array of detected keywords with their type (class, type, etc).

It will greatly simplify my coding of my editor syntax coloring. Also, it will remove duplication code, because I need to do it as a side code.

Regards,

**Comment by andreineculau** [comment]
Is this a pegjs issue (bug/feature)? If not, it's more appropriate to close this and post the question to https://groups.google.com/forum/#!forum/pegjs

**Comment by sguimont** [comment]
Sorry, if it was unclear, but it's a feature request.

I currently use peg.js for my DSL implementation and it works great!  But I really need to have the parsed language (tokens with their types)  and not the parsed result generated by the parsers.

I didn't find any api or plugins that provides me that, so I thing it will be great to have another method in peg.js parser to returns the list of tokens with their types.

**Comment by andreineculau** [comment]
Can you exemplify? What do you mean by "list of tokens with their types"?

I guess you're not talking about this: given

```
all = (keyword / number)+
keyword = "foo" / "bar"
number = [0-9]+
` ``

and parsing "for123bar" should say

```
[["for", "keyword"], ["123", "number"], ["bar", "keyword"]]
```

On Thu, Jan 22, 2015 at 3:42 PM, Sebastien Guimont <notifications@github.com
> wrote:

> Sorry, if it was unclear, but it's a feature request.
>
> I currently use peg.js for my DSL implementation and it works great! But I
> really need to have the parsed language (tokens with their types) and not
> the parsed result generated by the parsers.
>
> I didn't find any api or plugins that provides me that, so I thing it will
> be great to have another method in peg.js parser to returns the list of
> tokens with their types.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pegjs/pegjs/issues/311#issuecomment-71029760>.
>



-- 
andreineculau.com <http://www.andreineculau.com>

**Comment by sguimont** [comment]
This is exactly what I try to get.

I'm aware a token type is more complex than just a name, the same information as in SyntaxError.expected will be great.

```
all = (keyword / number)+
keyword = "foo" / "bar"
number = "number" [0-9]+
```

```
var tokens = parser.tokens("for123bar");
[
  {token: 'foo', type: 'literal', value: 'for', description: '"foo"' }, 
  {token: '123', type: 'other', description: ''number" }, 
  {token: 'bar', type: 'literal', value: 'bar', description: '"bar"' }
]
```

**Comment by sguimont** [comment]
I would like to know, if this feature request will be accepted or not please.  
If not, can you give me a hint on how I can get it?

**Comment by dmajda** [comment]
@sguimont Do I understand you correctly that you have an existing PEG.js grammar, you’d like to keep it mostly as it is (returning some kind of result that’s useful to you), but at the same time be able to use it to just tokenize the input and return the list of tokens?

**Comment by sguimont** [comment]
Yes, exactly.  Thanks. It will greatly help me to implements auto-completion and syntax coloring for my DSL.

**Comment by sguimont** [comment]
I just make a little follow-up.... @dmajda do you think it will be possible to add this feature to generated parser or maybe someone already provided a similar tool to get this the list of tokens from peg rules ?

**Comment by dmajda** [comment]
@sguimont Sorry I didn’t get back to you. You are far from the only person requesting this, so I’m marking it as a post-1.0.0 feature. I’m not aware of any external tool implementing this.

The main issue is how to distinguish token-level rules from high-level syntactic rules from and low-level rules that but themselves do not constitute tokens. Any ideas on how to do it conveniently are welcome. Note this is somewhat related to #11 which would need such classification too.

**Comment by StoneCypher** [comment]
`type` is a reserved word and will cause problems with legacy browsers and many approaches to attempting `typescript` support.  Even though it's strictly legal, I recommend that a different term be used, largely for the same reason that `react` has objects with `.className` instead of `.class` for CSS

I'd like to find out whether I understand this correctly.  I don't think that I do.

Given an input text and a parser grammar, produce a flat array of objects, where:

1. `token` is the actual text being represented;
    1. if you concatenate all `token`s in order, you resurrect the source text;
1. `type` is 
    1. `literal`, if what was matched was a string constant in the parser, and 
    1. `other`, otherwise;
1. `value` is `token` again, and only present on `literal`s;
1. `description` is 
    1. `token` again, but only on `literal`s, or
    1. the name of the rule, for `other`s

The shape of that is weird.  Is that the thing a syntax highlighter actually eats?

Are there things other than `literal` and `other`?  When I wrote a sublime highlighter there were like a dozen of them.  If so, what do you need, and how are they told apart?

----

As far as how to distinguish token and high level rules, I've never really been very clear on what a token actually is in a parser.  I kind of think it's the leaf node in the parse tree?

I think probably the right thing to do is to not actually decide this.  

The thing is, if what's being said here is "how do you provide a good choice of depth to represent as the leaf nodes to the tokenizer," meaning if I have three adjacent ten character strings, do I present one, three, or 30 tokens?

There are kind of a bunch of ways to go about this, and the right one isn't the same from parser to parser.  Honestly, the right thing to do isn't necessarily the same thing from the same parser from consumer to consumer.

Consider the case of a simple programming language implemented in `peg`.  And, because wolves are howling and the hour is late, let's say that it has some notation that resembles C-style macros.

1. For a linter, it's enough to detect that a block like that is in use.
1. For a programmer's editor, you'll want to be able to color the left cyan, and the right a different color for if it's public or private.  Or ... I don't know.  Also a third color for the actual hash signs.

Alternately, consider peg-driven text parsing tools for a logger.

1. For an error highlighter in a stream logger, it's enough to detect anything that doesn't match a known rule, as well as things that match known error rules.  
1. But for something that wants to actually parse the things coming over the stream, you might need a better granularity.

Fundamentally, these differences are small differences that should not reflect even a difference in the relevant grammars, much less in `peg` itself.

Therefore, this decision needs to be outside of `peg` and the grammars.

It's like deciding how best to implement `.sort()`.  

The short answer is "don't.  Implement a few common sense sorts, like locale-aware, locale-unaware, and efficient bytewise, then allow the user to provide a predicate to supply the alternatives."

Similarly, I believe that `peg` should:

1. Ignore this in general, for backwards compatability's sake
1. If you want this, you should pass a pair of functions in on `options`
    1. The first function is deep mapped over the AST, post-parse, and annotates the AST with its own result , which is an array of the ostensibly appropriate tokens at this location (eg `['literal','number']`)
        1. At a node `x`, the first function will need to be able to see neighbors sometimes, so a dom-ish API may be appropriate.  
        1. All the same, this doesn't modify the AST in any way other than to add the new value; they can only annotate their own node, not neighbors, parents, or children
    1. The second function takes the post-map decorated parse tree and reduces it to a token stream
        1. As long as it accepts an AST and returns something, his can be p.much whatever you want

Once you have that, there's no longer a need to search for "the right way" to do things.  We can just provide a couple common sense heuristics, like "if it only aggregates no-switch string literals and so do its neighbors, then you can use the parent node instead" or whatever

If the first function is an array of functions, instead of a function, then the AST token nodes will contain arrays of arrays-of-tokens, rather than arrays-of-tokens.  In that sense, a tree could be annotated multiple different ways by differing heuristics in a single pass.

And then the problem just boils down to offering some good baseline heuristics, giving those heuristics sufficiently clear names, and providing new interesting ones as people come up with them
